exec/java-exec/src/main/java/org/apache/parquet/hadoop/ParquetColumnChunkPageWriteStore.java

```diff
@@ -0,0 +1,269 @@
+
+  private final Map<ColumnDescriptor, ColumnChunkPageWriter> writers = Maps.newHashMap();
+  private final MessageType schema;
+
+  public ParquetColumnChunkPageWriteStore(BytesCompressor compressor,
+                                          MessageType schema,
+                                          int initialSlabSize,
+                                          int maxCapacityHint,
+                                          ByteBufferAllocator allocator) {
+    this.schema = schema;
+    for (ColumnDescriptor path : schema.getColumns()) {
+      writers.put(path,  new ColumnChunkPageWriter(path, compressor, initialSlabSize, maxCapacityHint, allocator));
+    }
+  }
+
+  @Override
+  public PageWriter getPageWriter(ColumnDescriptor path) {
+    return writers.get(path);
+  }
+
+  public void flushToFileWriter(ParquetFileWriter writer) throws IOException {
+    for (ColumnDescriptor path : schema.getColumns()) {
+      ColumnChunkPageWriter pageWriter = writers.get(path);
+      pageWriter.writeToFileWriter(writer);
+    }
+  }
+
+  @Override
+  public void close() {
+    for (ColumnChunkPageWriter pageWriter : writers.values()) {
+      pageWriter.close();
+    }
+  }
+
+  private static final class ColumnChunkPageWriter implements PageWriter, Closeable {
+
+    private final ColumnDescriptor path;
+    private final BytesCompressor compressor;
+
+    private final CapacityByteArrayOutputStream buf;
+    private DictionaryPage dictionaryPage;
+
+    private long uncompressedLength;
+    private long compressedLength;
+    private long totalValueCount;
+    private int pageCount;
+
+    // repetition and definition level encodings are used only for v1 pages and don't change
+    private Set<Encoding> rlEncodings = Sets.newHashSet();
+    private Set<Encoding> dlEncodings = Sets.newHashSet();
+    private List<Encoding> dataEncodings = Lists.newArrayList();
+
+    private Statistics totalStatistics;
+
+    private ColumnChunkPageWriter(ColumnDescriptor path,
+                                  BytesCompressor compressor,
+                                  int initialSlabSize,
+                                  int maxCapacityHint,
+                                  ByteBufferAllocator allocator) {
+      this.path = path;
+      this.compressor = compressor;
+      this.buf = new CapacityByteArrayOutputStream(initialSlabSize, maxCapacityHint, allocator);
+      this.totalStatistics = getStatsBasedOnType(this.path.getType());
+    }
+
+    @Override
+    public void writePage(BytesInput bytes,
+                          int valueCount,
+                          Statistics statistics,
+                          Encoding rlEncoding,
+                          Encoding dlEncoding,
+                          Encoding valuesEncoding) throws IOException {
+      long uncompressedSize = bytes.size();
+      if (uncompressedSize > Integer.MAX_VALUE) {
```

Comment:
If the uncompressed size is anywhere near `Integer.MAX_VALUE` (2 GB), then we have a resource problem. Double that for the compressed version and we're talking about serious memory usage. And, this is per column.

What is needed to ensure that each page (see [this diagram](https://parquet.apache.org/documentation/latest/)) is kept to a reasonable size? Does Drill already chose a reasonable size? Is it dependent on the (widely varying) input batch size?